pydantic/_internal/_model_construction.py

```diff
@@ -266,9 +267,31 @@ def build_inner_schema(  # noqa: C901
         cls.model_config['arbitrary_types_allowed'],
         local_ns,
     )
+
     return schema, fields
 
 
+def _simplify_recursive_schema(schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
+    """
+    Walk the schema recursively and any time a reference is repeated, replace it with a recursive-ref
+    """
+    visited: set[str] = set()
+
+    def replace_duplicate_refs(s: Any) -> Any:
+        if isinstance(s, dict):
+            if 'ref' in s and s.get('type') == 'typed-dict':
+                if s['ref'] in visited:
+                    return {'schema_ref': s['ref'], 'type': 'recursive-ref'}
+                else:
+                    visited.add(s['ref'])
+            return {k: replace_duplicate_refs(v) for k, v in s.items()}
+        elif isinstance(s, list):
+            return [replace_duplicate_refs(v) for v in s]
+        return s
```

Comment:
This is a pretty awful hack, and I think a better approach would be to rework the original schema generation to not produce "duplicate" refs. This would probably amount to replacing some of the recursive function calls (to `build_inner_schema` in particular) with a generator class (similar to the GenerateJsonSchema one) that would keep track of the references as it goes (rather than relying on the local/global_ns stuff).

Given that, I'm thinking we probably shouldn't merge this PR as is (in particular, because this `_simplify_recursive_schema` function is pretty unsafe), and should probably do a bigger refactor. But I wanted to push the commit with this function to demonstrate the issue more concretely.

I would believe it is possible to resolve this issue with some smart changes to the `build_inner_schema` and related functions, but @samuelcolvin I'm curious if you think it might work better to rework the schema generation into a class that would retain the references in an attribute? I think it might not actually be too much work, but I'd like to get your sign off before going down that path.

Reply:
Actually, I think this function may be useful, or even necessary, to handle a more core-schema-based approach to generics. Because if I want to perform a core-schema-substitution, and the substituted value has a ref in it, it might be good to remove all but one of the schemas with the same ref. I think I'll  make a more type-safe version of this function that has the sole purpose of replacing duplicate refs with recursive refs, and open a PR to pydantic-core to add it there. (If rejected there we can revisit this.)